core/src/avm2.rs

```diff
@@ -137,15 +144,20 @@ impl<'gc> Avm2<'gc> {
 
         let (method, scope, _domain) = script.init();
         match method {
-            Method::Native(method) => {
+            Method::Native(method, _) => {
                 //This exists purely to check if the builtin is OK with being called with
                 //no parameters.
                 init_activation.resolve_parameters(method.name, &[], &method.signature)?;
-
-                (method.method)(&mut init_activation, Some(scope), &[])?;
+                init_activation.context.avm2.push_global_init()?;
+                let r = (method.method)(&mut init_activation, Some(scope), &[]);
+                init_activation.context.avm2.pop_call();
+                r?;
             }
-            Method::Bytecode(_) => {
-                init_activation.run_stack_frame_for_script(script)?;
+            Method::Bytecode(_, _) => {
+                init_activation.context.avm2.push_global_init()?;
+                let r = init_activation.run_stack_frame_for_script(script);
+                init_activation.context.avm2.pop_call();
```

Comment:
personally not a fan of manual push/pop management; we might get an issue where we "catch" a `?` in Rust () but forget to pop the stack, right?
That said, not sure if there are better ways right now... plus, I imagine this would be refactored again for actual `throw` functionality anyway, right?

Reply:
What you're describing already happens with AVM1; `Avm1::Activation` actually stores the current call stack metadata in an on-stack linked list via `ActivationIdentifier`. The advantage of this is that you *can't* forget to pop something; the AVM and Rust stacks are linked together via borrow & drop checking. OTOH this design isn't really proven for error reporting, though - AVM1 doesn't really have runtime errors and `ActionThrow` is rarely used.

In that design, non-AVM activations are still expected to provide valid metadata; so for example if you want to set flashvars you need to identify yourself as `[Setting FlashVars]` or something. If we were to do something similar in AVM2 then we also have to create metadata for anything that might touch that VM.

The biggest problem would be that internal calls wouldn't show up on the stack trace - i.e. if one method of a function called another with the same `activation`. I don't know if this is actually a problem, though.

Reply:
My understanding is that AVM2 has a very C-like stack, that... literally works like C(++)? It stores stack variables, arguments, function pointers that are used for unwinding etc. 
So in principle the "activation chain" should do similar job, except I don't think anyone thought about actual exception handling yet, so it's not guaranteed it won't need a rewrite for that then.

> In that design, non-AVM activations are still expected to provide valid metadata; so for example if you want to set flashvars you need to identify yourself as [Setting FlashVars] or something. If we were to do something similar in AVM2 then we also have to create metadata for anything that might touch that VM.

Yeah, but we already do this, right?

> The biggest problem would be that internal calls wouldn't show up on the stack trace - i.e. if one method of a function called another with the same activation. I don't know if this is actually a problem, though.

For the sake of dev debugging, as long as we have access to both Rust and Flash tracebacks, we should be fine. In fact, if we wanted to be fancy, we could even mix them into one :)

(But if we wanted to stay with separate call stack, let's at least make convenience methods that wrap `.push(); call(); .pop();` so that you never need to do that manually.)